lib/jekyll/collection.rb

```diff
@@ -210,6 +210,65 @@ def read_document(full_path)
 
     private
 
+    def sort_docs!
+      if @metadata["order"] && @metadata["order"].is_a?(Array)
+        rearrange_docs!
+      elsif @metadata["sort_by"]
+        sort_docs_by_key!
+      else
+        docs.sort!
+      end
+    end
+
+    def sort_docs_by_key!
+      meta_key = @metadata["sort_by"]
+      docs.sort_by! { |d| d.data[meta_key] }
+    rescue
+      # Inform which document doesn't have the required key defined, and proceed
+      # with the default sort function.
+      #
+      # But if the user defines the key with a different Class type, (e.g. Integer
+      # instead of a String), simply skip custom sorting, and proceed.
+      docs.select { |entry| entry.data[meta_key].nil? }.each do |e|
+        Jekyll.logger.warn "Error:", "'#{meta_key}' not defined in '#{e.relative_path}'"
+      end
+
+      Jekyll.logger.warn "Error:",
+        "Custom sorting skipped due to inconsistent key definition."
+    end
+
+    def rearrange_docs!
+      @metadata["order"].each do |entry|
+        doc_path = "_#{@label}/#{entry}"
+        doc = docs.find { |d| d.relative_path == doc_path }
+
+        if File.exist?(@site.in_source_dir(doc_path))
+          doc.data["sort_index"] = @metadata["order"].index(entry)
```

Comment:
I don't want us to be munging the data for the document. Couldn't this method be simplified to just:

```ruby
docs.sort_by! { |doc| metadata["order"].index(doc.relative_path.sub(relative_directory, "")) || docs.size }
```